crates/core_simd/examples/README.md

```diff
@@ -0,0 +1,109 @@
+
+We can observe a couple of interesting things 
+1. Both codes output almost identical assembly if the `opt-level` is set to `3`. What happens if you set them to `0,1,2`? Do they always give the same stats?
+2. Try disabling the `assert!` comment. What happens to the assembly? Why do you think that people avoid `panic!` in tight Rust code?
+3. You can target different instruction sets that have shorter/wider SIMD lanes. Trying adding `-target-feature=+avx2` in the compilation options (next to the `opt-level` flags) and see how the assembly changes.
+4. Just because `xmm` registers are being used, doesn't necessarily mean you're getting SIMD speedups. This is in part due to technical debt and backwards compatibility that `xmm` registers are easier to pass floats around in, but to really get them to operate on multiple data we need more direct control that what we are currently using.
+5. TODO In order to maximize throughput of our tight loop, we are looking to maximize/minimize these following measurements: throughput, ipc/cycle, etc. 
+
+-----
+
+#### SIMD version of `dot_product`
+
+Open up [this link in your browser of choice](https://rust.godbolt.org/z/85neY7Kcn). For the code on the left, the full suite of optimizations kick in when we get to `opt-level=2`, and we get ~50 lines of assembly total. When starting out, more compact assembly can be a decent indication of more streamlined coding, but it quickly dies out as a heuristic, so don't get too distracted by that factor. Looking at the `llvm-mca` window on the bottom left, we observe we get an (TODO IPC/throughput) of XXX, whereas in the previous example, our best code could only get to YYY. A couple of notes on this snippet:
+
+```rust
+#![feature(portable_simd)]
+#![feature(array_chunks)]
+use std::simd::*;
+
+// Other options to try instead of "avx2": 
+// "sse"
+// "sse4.1"
+//#[target_feature(enable = "avx2")]
+pub unsafe fn dot_prod_simd_0(a: &[f32], b: &[f32]) -> f32 {
+    // TODO handle remainder when a.len() % 4 != 0
+    a.array_chunks::<4>()
+        .map(|&a| f32x4::from_array(a))
+        .zip(b.array_chunks::<4>().map(|&b| f32x4::from_array(b)))
+        .map(|(a, b)| (a * b).reduce_sum())
+        .sum()
+
+```
+
+1. SIMD comes in many flavors (instructions sets). These (like `sse`, `sse4.1`, `avx2`) describe the hardware capabilities of your current CPU. That is, if you don't have `avx512`, you physically do not have a SIMD vector that can hold 512 bytes at a time at most on your CPU. 
+2. You can switch between different instruction sets by changing the `#![target-feature(...)]` macro above the function, as well as declaring it unsafe.
+3. Inside Godbolt, you can hover over an instruction to display a tooltip of what it says. Try hovering your mouse over `mulps` and reading what it says.
```

Comment:
I suggest phrasing in terms of "what the instruction does" rather than "what it says".